src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs

```diff
@@ -38,35 +39,40 @@ public Task<int> ReadAsync(ArraySegment<byte> buffer, CancellationToken cancella
             return result;
         }
 
-        public async Task Consume(CancellationToken cancellationToken = default(CancellationToken))
+        public Task Consume(CancellationToken cancellationToken = default(CancellationToken))
         {
-            Task<int> result;
-            var send100checked = false;
-            do
+            var result = ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken);
+            if (!result.IsCompleted)
             {
-                result = ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken);
-                if (!result.IsCompleted)
-                {
-                    if (!send100checked)
-                    {
-                        if (Interlocked.Exchange(ref _send100Continue, 0) == 1)
-                        {
-                            _context.FrameControl.ProduceContinue();
-                        }
-                        send100checked = true;
-                    }
-                }
-                else if (result.GetAwaiter().GetResult() == 0) 
-                {
-                    // Completed Task, end of stream
-                    return;
-                }
-                else
+                if (Interlocked.Exchange(ref _send100Continue, 0) == 1)
                 {
-                    // Completed Task, get next Task rather than await
-                    continue;
+                    _context.FrameControl.ProduceContinue();
                 }
-            } while (await result != 0);
+
+                return ConsumeAwaited(result, cancellationToken);
+            }
+            else if (result.GetAwaiter().GetResult() == 0)
+            {
+                // Completed Task, end of stream
+                return TaskUtilities.CompletedTask;
+            }
+            else
+            {
+                // Completed Task, but non-zero get next Task and await
+                return ConsumeAwaited(ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken), cancellationToken);
```

Comment:
I don't understand this. Why not put his call to `ReadAsyncImplementation` inside of `ConsumeAwaited`? You already do this inside the while condition.